File: game-server/src/main/java/net/dodian/uber/game/runtime/net/OutboundSessionQueue.java
package net.dodian.uber.game.runtime.net;

import io.netty.channel.Channel;
import java.util.ArrayDeque;
import net.dodian.uber.game.netty.codec.ByteMessage;

/**
 * Per-client queued outbound transport. Messages are appended in call order and
 * drained once per tick.
 */
public final class OutboundSessionQueue {

    public static final class DrainResult {
        private final int messageCount;
        private final int byteCount;

        private DrainResult(int messageCount, int byteCount) {
            this.messageCount = messageCount;
            this.byteCount = byteCount;
        }

        public int messageCount() {
            return messageCount;
        }

        public int byteCount() {
            return byteCount;
        }

        public static DrainResult empty() {
            return new DrainResult(0, 0);
        }
    }

    private final ArrayDeque<ByteMessage> queuedMessages = new ArrayDeque<>();

    public synchronized void enqueue(ByteMessage message) {
        queuedMessages.addLast(message);
    }

    public synchronized boolean isEmpty() {
        return queuedMessages.isEmpty();
    }

    public synchronized DrainResult drainTo(Channel channel) {
        if (queuedMessages.isEmpty()) {
            return DrainResult.empty();
        }
        int messages = 0;
        int bytes = 0;
        while (!queuedMessages.isEmpty()) {
            ByteMessage message = queuedMessages.removeFirst();
            bytes += Math.max(0, message.content().writerIndex());
            channel.write(message);
            messages++;
        }
        return new DrainResult(messages, bytes);
    }

    public synchronized void releaseAll() {
        while (!queuedMessages.isEmpty()) {
            queuedMessages.removeFirst().releaseAll();
        }
    }
}

File: game-server/src/main/kotlin/net/dodian/uber/game/runtime/net/OutboundSessionQueue.kt
package net.dodian.uber.game.runtime.net

import io.netty.channel.Channel
import java.util.ArrayDeque
import net.dodian.uber.game.netty.codec.ByteMessage

/**
 * Per-client queued outbound transport. Messages are appended in call order and
 * drained once per tick.
 */
class OutboundSessionQueue {
    class DrainResult private constructor(
        private val messageCount: Int,
        private val byteCount: Int,
    ) {
        fun messageCount(): Int = messageCount

        fun byteCount(): Int = byteCount

        companion object {
            private val EMPTY = DrainResult(0, 0)

            @JvmStatic
            fun empty(): DrainResult = EMPTY

            @JvmStatic
            fun of(messageCount: Int, byteCount: Int): DrainResult =
                if (messageCount == 0 && byteCount == 0) {
                    EMPTY
                } else {
                    DrainResult(messageCount, byteCount)
                }
        }
    }

    private val queuedMessages = ArrayDeque<ByteMessage>()

    @Synchronized
    fun enqueue(message: ByteMessage) {
        queuedMessages.addLast(message)
    }

    @Synchronized
    fun isEmpty(): Boolean = queuedMessages.isEmpty()

    @Synchronized
    fun drainTo(channel: Channel): DrainResult {
        if (queuedMessages.isEmpty()) {
            return DrainResult.empty()
        }
        var messages = 0
        var bytes = 0
        while (!queuedMessages.isEmpty()) {
            val message = queuedMessages.removeFirst()
            bytes += maxOf(0, message.content().writerIndex())
            channel.write(message)
            messages++
        }
        return DrainResult.of(messages, bytes)
    }

    @Synchronized
    fun releaseAll() {
        while (!queuedMessages.isEmpty()) {
            queuedMessages.removeFirst().releaseAll()
        }
    }
}
